File: 2025/9/9_26_leetcode_bc/105.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (preorder.length === 0) return null;

    const root = new TreeNode(preorder[0]);

    const stack = [root];
    let i = 0;

    for (const val of preorder.slice(1)) {
        let node = stack[stack.length - 1];

        if (node.val !== inorder[i]) {
            node.left = new TreeNode(val);
            stack.push(node.left)
        } else {
            while (stack.length && stack[stack.length - 1].val === inorder[i]) {
                node = stack.pop();
                i += 1;
            }
            node.right = new TreeNode(val);
            stack.push(node.right);
        }
    }

    return root;
};

File: 2025/9/9_26_leetcode_bc/105.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None

        pos = {v: i for i, v in enumerate(inorder)}

        def build(l:int, h:int, pre_index:int):
            if l > h or pre_index >= len(preorder):
                return None, pre_index

            cur_val = preorder[pre_index]
            cur_root = TreeNode(cur_val)
            mid_index = pos[cur_val]

            cur_root.left, next_pre_index = build(l, mid_index-1, pre_index+1)
            cur_root.right, next_pre_index = build(mid_index+1, h, next_pre_index)

            return cur_root, next_pre_index

        root, _ = build(0, len(preorder), 0)
        return root
